openlibrary/plugins/worksearch/code.py

```diff
@@ -1192,15 +1192,18 @@ def rewrite_list_editions_query(q, page, offset, limit):
     return the query, unchanged, exactly as it entered the
     function. If it does contain a lists key, then use the pagination
     information to fetch the right block of keys from the
-    lists_editions API and then feed these editions resulting work
+    lists_editions and lists_works API and then feed these editions resulting work
     keys into solr with the form key:(OL123W, OL234W). This way, we
     can use the solr API to fetch list works and render them in
     carousels in the right format.
     """
     if '/lists/' in q:
         editions = get_list_editions(q, offset=offset, limit=limit)
-        work_ids = [ed.get('works')[0]['key'] for ed in editions]
-        q = 'key:(' + ' OR '.join(work_ids) + ')'
+        works = get_list_works(q, offset=offset, limit=limit)
```

Comment:
This may not get us exactly what we want because it paginates over both editions and works and makes two separate calls.

We probably want a single call called `get_list_books` which makes a single call to what you're currently calling `get_works` (which might be more appropriately called `get_books` which selects anything which is either `/type/work` **or** `/type/edition`).

Pagination may have to be tweaked a little bit so that we first fetch all of the list items, then filter based on which ones are books (i.e. editions or works) and then finally paginate the resulting list.

Reply:
Okay, so I'll create two new functions, get_list_books and get_books to replace get_list_works and get_works and fetch both editions and works using those functions. Do I have make any extra changes to pagination or should making the above mentioned changes be sufficient @mekarpeles ??

Reply:
@b2pacific we're keeping your change and making some updates so you get credit for the PR! @jimchamp and I were able to leverage your code in simplifying the solution. Thank you!!